Source/Particles/Deposition/ChargeDeposition.H

```diff
@@ -201,6 +201,322 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
+    if( load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::GpuClock) {
+        cost_real = (amrex::Real *) amrex::The_Managed_Arena()->alloc(sizeof(amrex::Real));
+        *cost_real = 0.;
+    }
+#endif
+
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+    const auto dxiarr = geom.InvCellSizeArray();
+    const auto plo = geom.ProbLoArray();
+    const auto domain = geom.Domain();
+
+    const amrex::Box sample_tbox(IntVect(AMREX_D_DECL(0, 0, 0)), a_tbox_max_size - 1);
+    amrex::Box sample_tbox_x = convert(sample_tbox, ix_type);
+
+    sample_tbox_x.grow(depos_order);
+
+    const auto npts = sample_tbox_x.numPts();
+
+    const int nblocks = a_bins.numBins();
+    const auto offsets_ptr = a_bins.offsetsPtr();
+    const int threads_per_block = 256;
+
+    std::size_t shared_mem_bytes = npts*sizeof(amrex::Real);
+    amrex::IntVect bin_size = WarpX::sort_bin_size;
+
+    const std::size_t max_shared_mem_bytes = amrex::Gpu::Device::sharedMemPerBlock();
+
+    if (shared_mem_bytes > max_shared_mem_bytes) {
+      std::cout << "Tile size for charge deposition of " << shared_mem_bytes
+                << " bytes would not fit in GPU's shared memory of " << max_shared_mem_bytes << " bytes" << std::endl;
+      amrex::Abort("Tile size too big for GPU shared memory charge deposition");
+    }
+    ignore_unused(np_to_deposit);
+#endif
+
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+    amrex::launch(
+                  nblocks, threads_per_block, shared_mem_bytes, amrex::Gpu::gpuStream(),
+                  [=] AMREX_GPU_DEVICE () noexcept
+#else // defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+    amrex::ParallelFor(np_to_deposit, [=] AMREX_GPU_DEVICE (long ip_orig) noexcept
+#endif
+      {
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+        const int bin_id = blockIdx.x;
+        const unsigned int bin_start = offsets_ptr[bin_id];
+        const unsigned int bin_stop = offsets_ptr[bin_id+1];
+
+        if (bin_start == bin_stop) { return; }
+
+        amrex::Box buffer_box;
+        {
+          ParticleReal xp, yp, zp;
+          GetPosition(permutation[bin_start], xp, yp, zp);
+#if defined(WARPX_DIM_3D)
+          IntVect iv = IntVect(int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((yp-plo[1])*dxiarr[1])),
+                               int(amrex::Math::floor((zp-plo[2])*dxiarr[2])));
+#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+          IntVect iv = IntVect(
+                               int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((zp-plo[1])*dxiarr[1])));
+#elif defined(WARPX_DIM_1D_Z)
+          IntVect iv = IntVect(int(amrex::Math::floor((zp-plo[0])*dxiarr[0])));
+#endif
+          iv += domain.smallEnd();
+          getTileIndex(iv, box, true, bin_size, buffer_box);
+        }
+
+        Box tbx = convert( buffer_box, ix_type);
+        tbx.grow(depos_order);
```

Comment:
Hm, this would imply that we need 3 guard cells for `depos_order = 3`, which does not seem correct.
Do we think that we could reduce this? (It matters for shared memory usage, esp. for 3D)

Reply:
Are you thinking it's simply one larger than it needs to be - i.e. that `depos_order = 1` should need no additional guard cells, and go from there?

Otherwise, this has more subtlety than I can confidently answer.